### PR TITLE
Rewrite `discouraged_optional_collection` rule with SwiftSyntax

### PR DESCRIPTION
**Pull request by woxtu**
Hi,

This PR rewrites `discouraged_optional_collection` rule with SwiftSyntax and match the behavior of the rule to `discouraged_optional_boolean` rule.

### CODE REVIEW

**Comment by SimplyDanny** [comment]
Hey @woxtu, thank you for all your contributions so far. I currently don't have access to my notebook and also miss the time to review your PRs. This will change in about a week though. So please bear with us a little while longer. You'll receive feedback in some days.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionRule.swift [hunk 1]
```diff
@@ -12,153 +12,66 @@ struct DiscouragedOptionalCollectionRule: ASTRule, OptInRule, ConfigurationProvi
...
-                range.upperBound < endIndex,
-                let initialIndex = index(range.lowerBound, offsetBy: -3, limitedBy: startIndex),
-                let finalIndex = index(range.upperBound, offsetBy: 1, limitedBy: endIndex),
-                self[initialIndex..<range.lowerBound] == "Set",
-                self[range.upperBound] == "?" else { return nil }
+                let calledExpression = node.calledExpression.as(MemberAccessExprSyntax.self),
+                let base = calledExpression.base?.as(IdentifierExprSyntax.self),
+                base.identifier.text == "Optional",
+                calledExpression.name.text == "some",
+                let expression = node.argumentList.first?.expression
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionRule.swift:40]: Honestly, I don't expect people to ever right an optional collection in this way, but it's already there so we should keep it.

Later we can check if removing this case causes any regressions in OSS projects.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionRule.swift [hunk 2]
```diff
@@ -12,153 +12,66 @@ struct DiscouragedOptionalCollectionRule: ASTRule, OptInRule, ConfigurationProvi
...
+            else {
+                return
+            }
 
-            return initialIndex..<finalIndex
+            if expression.isCollectionExpression {
+                violations.append(node.positionAfterSkippingLeadingTrivia)
```
[the remark this answers is not in the document]
Comment (woxtu) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionRule.swift:46]: Yes.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionExamples.swift [hunk 3]
```diff
@@ -56,155 +56,163 @@ internal struct DiscouragedOptionalCollectionExamples {
...
+        wrapExample("enum", "static func foo(input: ↓[String: [String: String]]?) {}"),
+        wrapExample("enum", "static func foo(input: [String: ↓[String: String]?]) {}"),
+        wrapExample("enum", "static func foo(input: ↓[String: ↓[String: String]?]?) {}"),
+        wrapExample("enum", "static func foo<K, V>(_ dict1: [K: V], _ dict2: ↓[K: V]?) -> [K: V]"),
+        wrapExample("enum", "static func foo<K, V>(dict1: [K: V], dict2: ↓[K: V]?) -> [K: V]"),
+
+        // Optional chaining
+        Example("_ = ↓[Int]?.values()"),
+        Example("_ = ↓[String: String]?.values()"),
+        Example("_ = ↓Set<Int>?.values()")
```
[the remark this answers is not in the document]
Comment (woxtu) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionExamples.swift:215]:
```swift
extension Optional where Wrapped == [Int] {
    static func values() {}
}

[Int]?.values() // ok
```

I think no one do this but `discouraged_optional_boolean` rule checks `Bool?.values()`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,10 @@
   (e.g. `@testable`, `@_exported`, ...).  
   [hiltonc](https://github.com/hiltonc)
 
+* Rewrite `discouraged_optional_collection` rule with SwiftSyntax and match
+  the behavior of the rule to `discouraged_optional_boolean` rule.
```
[the remark this answers is not in the document]
Comment (woxtu) [reply CHANGELOG.md:29]: I'll merge this later.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionExamples.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionExamples.swift:215]: This looks very strange. 🤔 But fair enough. Please keep it. It's valid code indeed.

**Comment by SimplyDanny** [comment]
The rule has meanwhile been updated in another PR. Thanks for all the work nevertheless, @woxtu!

**Pull request closed without merging**